libraries/chain/include/eosio/chain/webassembly/interface.hpp

```diff
@@ -0,0 +1,581 @@
+         REGISTER_LEGACY_HOST_FUNCTION(check_permission_authorization);
+         REGISTER_HOST_FUNCTION(get_permission_last_used);
+         REGISTER_HOST_FUNCTION(get_account_creation_time);
+
+         // authorization api
+         void require_auth(account_name);
+         void require_auth2(account_name, permission_name);
+         bool has_auth(account_name) const;
+         void require_recipient(account_name);
+         bool is_account(account_name) const;
+
+         REGISTER_HOST_FUNCTION(require_auth);
+         REGISTER_HOST_FUNCTION(require_auth2);
+         REGISTER_HOST_FUNCTION(has_auth);
+         REGISTER_HOST_FUNCTION(require_recipient);
+         REGISTER_HOST_FUNCTION(is_account);
+
+         // system api
+         uint64_t current_time() const;
+         uint64_t publication_time() const;
+         bool is_feature_activated(const digest_type&) const;
+         name get_sender() const;
+
+         REGISTER_HOST_FUNCTION(current_time);
+         REGISTER_HOST_FUNCTION(publication_time);
+         REGISTER_HOST_FUNCTION(is_feature_activated);
+         REGISTER_HOST_FUNCTION(get_sender);
+
+         // context-free system api
+         void abort();
+         void eosio_assert(bool, null_terminated_ptr);
+         void eosio_assert_message(bool, legacy_array_ptr<const char>);
+         void eosio_assert_code(bool, uint64_t);
+         void eosio_exit(int32_t);
+
+         REGISTER_CF_HOST_FUNCTION(abort)
+         REGISTER_LEGACY_CF_HOST_FUNCTION(eosio_assert)
+         REGISTER_LEGACY_CF_HOST_FUNCTION(eosio_assert_message)
+         REGISTER_CF_HOST_FUNCTION(eosio_assert_code)
+         REGISTER_CF_HOST_FUNCTION(eosio_exit)
+
+         // action api
+         int32_t read_action_data(legacy_array_ptr<char>) const;
+         int32_t action_data_size() const;
+         name current_receiver() const;
+         void set_action_return_value(legacy_array_ptr<char>);
+
+         REGISTER_LEGACY_CF_HOST_FUNCTION(read_action_data);
+         REGISTER_CF_HOST_FUNCTION(action_data_size);
+         REGISTER_CF_HOST_FUNCTION(current_receiver);
+         REGISTER_LEGACY_HOST_FUNCTION(set_action_return_value);
+
+         // console api
+         void prints(null_terminated_ptr);
+         void prints_l(legacy_array_ptr<const char>);
+         void printi(int64_t);
+         void printui(uint64_t);
+         void printi128(const __int128&);
+         void printui128(const unsigned __int128&);
+         void printsf(float);
+         void printdf(double);
+         void printqf(const float128_t&);
+         void printn(name);
+         void printhex(legacy_array_ptr<const char>);
+
+         REGISTER_LEGACY_CF_HOST_FUNCTION(prints);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(prints_l);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(printi);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(printui);
+         REGISTER_CF_HOST_FUNCTION(printi128);
+         REGISTER_CF_HOST_FUNCTION(printui128);
```

Comment:
Again, I think we will need to treat the reference types in `printi128` and `printui128` differently.